src/editor/codemirror/dnd.ts

```diff
@@ -0,0 +1,77 @@
+import { ChangeSet, Transaction } from "@codemirror/state";
+import { EditorView } from "@codemirror/view";
+import { pythonSnippetMediaType } from "../../common/mediaTypes";
+import { calculateChanges } from "./edits";
+
+// No access to the content via the event in dragover (as it may be cross-document)
+// so we need shared state with the drag so we can draw previews of the drop in
+// the document.
+let draggedCode: string | undefined;
+let invertedChanges: ChangeSet | undefined;
+
+export const setDraggedCode = (value: string | undefined) => {
+  draggedCode = value;
+};
```

Comment:
This kinda sucks, but I'm not sure it's worth the complication of avoiding the globals for this, particularly given it is part React and part CM. Leaving this open to ponder.